metadata/src/main/java/org/apache/kafka/metadata/bootstrap/BootstrapDirectory.java

```diff
@@ -0,0 +1,134 @@
+import org.apache.kafka.server.common.MetadataVersion;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static java.nio.file.StandardCopyOption.ATOMIC_MOVE;
+import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
+import static org.apache.kafka.server.common.MetadataVersion.MINIMUM_KRAFT_VERSION;
+
+
+/**
+ * A read-only class that holds the controller bootstrap metadata. A file named "bootstrap.snapshot" is used and the
+ * format is the same as a KRaft snapshot.
+ */
+public class BootstrapDirectory {
+    final static String INTER_BROKER_PROTOCOL_CONFIG_KEY = "inter.broker.protocol.version";
+    final static String BINARY_BOOTSTRAP = "bootstrap.checkpoint";
+
+    public static String ibpStringFromConfigMap(Map<String, Object> staticConfig) {
+        Object value = staticConfig.get(INTER_BROKER_PROTOCOL_CONFIG_KEY);
+        return value == null ? "" : value.toString();
+    }
+
+    private final String directoryPath;
+    private final String ibp;
+
+    /**
+     * Create a new BootstrapDirectory object.
+     *
+     * @param directoryPath     The path to the directory with the bootstrap file.
+     * @param ibp               The configured value of inter.broker.protocol, or the empty string
+     *                          if it is not configured.
+     */
+    public BootstrapDirectory(
+        String directoryPath,
+        String ibp
+    ) {
+        Objects.requireNonNull(directoryPath);
+        Objects.requireNonNull(ibp);
+        this.directoryPath = directoryPath;
+        this.ibp = ibp;
+    }
+
+    public BootstrapMetadata read() throws Exception {
+        if (!Files.isDirectory(Paths.get(directoryPath))) {
+            throw new RuntimeException("No such directory as " + directoryPath);
+        }
+        Path binaryBootstrapPath = Paths.get(directoryPath, BINARY_BOOTSTRAP);
+        if (!Files.exists(binaryBootstrapPath)) {
+            return readFromConfiguration();
+        } else {
+            return readFromBinaryFile(binaryBootstrapPath.toString());
+        }
+    }
+
+    BootstrapMetadata readFromConfiguration() {
+        if (ibp.isEmpty()) {
+            return BootstrapMetadata.fromVersion(MetadataVersion.latest(),
+                    "the default bootstrap file which sets the latest metadata.version, " +
+                    "since no bootstrap file was found, and " + INTER_BROKER_PROTOCOL_CONFIG_KEY +
+                    " was not configured.");
+        }
+        MetadataVersion version = MetadataVersion.fromVersionString(ibp);
+        if (version.isLessThan(MINIMUM_KRAFT_VERSION)) {
+            return BootstrapMetadata.fromVersion(MINIMUM_KRAFT_VERSION,
+                    "a default bootstrap file setting the minimum supported KRaft metadata " +
```

Comment:
Well, this is an exception that will terminate the startup process. Making it concise is probably not what we want.

"source" is also used in logging in the case where startup succeeded. It's useful to know things like "you configured X, but actually you got Y, because of these reasons" or "you didn't configure anything, so you got X, have a nice day"

Reply:
My point is that it does not read coherently when inserted into the exception message. For example:

> Unable to load metadata from a default bootstrap file setting the minimum supported KRaft metadata version, since no bootstrap file was found, and inter.broker.protocol.version was XX which is not supported by KRaft.: bootstrap metadata versions less than 3.3.IV1 are not supported.

Reply:
OK. I will reorganize this so that "source" is always at the end of the exception text, and make it a little bit shorter.

I still think we should err on the side of being a bit verbose and explaining exactly what is happening, even if it's a little janky. Having to guess about this stuff is very very bad. Often all you have to go on in these problems is an exception, log message, etc. so we should be sure that it actually does explain what happened.

Reply:
I don't mind having verbose messages, but I don't think it's asking too much for them to be grammatical.